dev_setup.sh

```diff
@@ -0,0 +1,155 @@
+            echo "Unsupported platform" 2>&1; exit 1;
+        fi;;
+    *) echo "Unsupported platform" 2>&1; exit 1;;
+esac
+
+function notice {
+    echo "[NOTICE] $*"
+}
+
+# Requires: LOG set to log file path.
+function logged {
+    if ! [ -z "$LOG" ]
+    then
+        # This may or may not be the same directory that this script lives in.
+        mkdir -p `dirname $LOG`
+        echo "$@" >>$LOG
+        if ! "$@" >>$LOG 2>&1
+        then
+            echo
+            echo "An error occurred; please see $LOG"
+            echo "Here are the last 50 lines:"
+            tail -n 50 $LOG
+            exit 1
+        fi
+    fi
+}
+
+function update_submodules {
+    cd $ROOT
+    notice "Updating submodules"
+    logged git submodule update --init
+}
+
+function install_ghcup {
+    if ! ghcup --version &> /dev/null
+    then
+        notice "Installing ghcup, GHC, and cabal"
+        # Technically the installation only requires cabal, but it's
+        # recommended to get the whole GCH shebang in one package.
+        # The output is not routed to log because the installer is interactive.
+        curl --proto '=https' --tlsv1.2 -sSf https://get-ghcup.haskell.org | sh
+    else
+        notice "Using existing ghcup installation"
+    fi
+
+}
+
+function install_gmp {
+    if [ $CRYPTOL_PLATFORM = $MACOS12 ] || [ $CRYPTOL_PLATFORM = $MACOS14 ]; then
+        notice "Installing GMP via Homebrew, if it's not already installed"
+        logged brew install gmp
+
+        # `brew --prefix` is different on macOS 12 and macOS 14
+        echo "export CPATH=$(brew --prefix)/include" >> $VAR_FILE
+        echo "export LIBRARY_PATH=$(brew --prefix)/lib" >> $VAR_FILE
+        notice "You may need to source new environment variables added here:\n" \
+            "\$ source $VAR_FILE"
+    else
+        notice "Did not install GMP. This script only supports macOS 14"
+    fi
+}
+
+# Installs the two solvers required to run the test suite for the repo.
+# Users may want to install other solvers, and indeed the what4 solvers repo
+# includes a half dozen other solvers that are compatible with cryptol. 
+function install_what4_solvers {
+    if ! cvc4 --version &> /dev/null || ! cvc5 --version &> /dev/null; then
+        notice "Installing cvc4 and/or cvc5 solvers"
+
+        if [ $CRYPTOL_PLATFORM = $MACOS14 ]; then
+            solvers_version=$WHAT4_SOLVERS_MACOS_14
+        elif [ $CRYPTOL_PLATFORM = $MACOS12 ]; then
+            solvers_version=$WHAT4_SOLVERS_MACOS_12
+        fi
+
+        solvers_dir=$(mktemp -d)
+        logged curl --proto '=https' --tlsv1.2 -sSfL -o "$solvers_dir/solvers.bin.zip" "$WHAT4_SOLVERS_URL$solvers_version"
+        cd $solvers_dir
+        logged unzip solvers.bin.zip
+        
+        # If we want to install more solvers by default, we can do so here,
+        # although we might need a different check than `--version`
+        for solver in cvc4 cvc5; do
+            if ! $solver --version &> /dev/null; then 
```

Comment:
Should this be `if ! ./$solver ...` since frequently `.` is explicitly not in PATH for security reasons?  This also helps ensure this operates on the just unpacked file and not something that might be in the PATH (which was verified in line 113, but is still a "bad smell").  Also, the notice in the following line should probably indicate where the solvers are being installed so the user can ensure they are in their path.  And I guess perhaps this whole `then` block might want to check if `/usr/local/bin` is in the path in case these are actually installed and it's just a path issue.

Reply:
Oh, this is confusing. This explicitly is trying to run the version on the path, _not_ the version in the local directory. The parent conditional says if we're missing either of these solvers, we need to download the whole set of binaries from what4_solvers. But if you're only missing one of them, we don't want to clobber the version you do have, so this double-checks for each solver whether it's already in the path and only installs it if it not.

I'm adding an `is_installed` method to replace this version check that will hopefully make it more obvious what's happening.